src/ordered_robin_dict.jl

```diff
@@ -0,0 +1,476 @@
+
+"""
+    empty!(collection) -> collection
+
+Remove all elements from a `collection`.
+
+# Examples
+```jldoctest
+julia> A = OrderedRobinDict("a" => 1, "b" => 2)
+OrderedRobinDict{String,Int64} with 2 entries:
+  "a" => 1
+  "b" => 2
+
+julia> empty!(A);
+
+julia> A
+OrderedRobinDict{String,Int64} with 0 entries
+```
+"""
+function empty!(h::OrderedRobinDict{K,V}) where {K, V}
+    empty!(h.dict)
+    empty!(h.keys)
+    empty!(h.vals)
+    h.count = 0
+    return h
+end
+
+function _setindex!(h::OrderedRobinDict, v, key)
+    hk, hv = h.keys, h.vals
+    push!(hk, key)
+    push!(hv, v)
+    nk = length(hk)
+    @inbounds h.dict[key] = Int32(nk)
+    h.count += 1
+end
+
+function setindex!(h::OrderedRobinDict{K, V}, v0, key0) where {K,V}
+    key = convert(K,key0)
+
+    v = convert(V,  v0)
+
+    index = get(h.dict, key, -2)
+
+    if index < 0
+        _setindex!(h, v0, key0)
+    else
+        @assert haskey(h, key0)
+        @inbounds orig_v = h.vals[index]
+        (orig_v != v0) && (@inbounds h.vals[index] = v0)
+    end
+
+    check_for_rehash(h) && rehash!(h)
+
+    return h
+end
+
+# rehash when there are ALLOWABLE_USELESS_GROWTH %
+# tombstones, or non-mirrored entries in the dictionary
+function check_for_rehash(h::OrderedRobinDict)
+    keysl = length(h.keys)
+    dictl = length(h)
+    return (keysl > (1 + ALLOWABLE_USELESS_GROWTH)*dictl)
+end
+
+function rehash!(h::OrderedRobinDict{K, V}) where {K, V}
+    keys = h.keys
+    vals = h.vals
+    newk = Vector{K}()
+    newv = Vector{V}()
+    hk, hv = newk, newv
+    
+    for (idx, (k, v)) in enumerate(zip(keys, vals))
+        if get(h.dict, k, -1) == idx
+            ccall(:jl_array_grow_end, Cvoid, (Any, UInt), hk, 1)
+            nk = length(hk)
+            @inbounds hk[nk] = k
+            ccall(:jl_array_grow_end, Cvoid, (Any, UInt), hv, 1)
+            @inbounds hv[nk] = v
+        end
+    end
+    
+    h.keys = newk
+    h.vals = newv
+    
+    for (idx, k) in enumerate(h.keys)
+        h.dict[k] = idx
+    end
+    return h
+end
+
+function sizehint!(d::OrderedRobinDict, newsz)
+    oldsz = length(d)
+    # grow at least 25%
+    if newsz < (oldsz*5)>>2
+        return d
+    end
+    sizehint!(d.keys, newsz)
+    sizehint!(d.vals, newsz)
+    sizehint!(d.dict, newsz)
+    return d
+end
+
+"""
+    get!(collection, key, default)
+
+Return the value stored for the given key, or if no mapping for the key is present, store
+`key => default`, and return `default`.
+
+# Examples
+```jldoctest
+julia> d = OrderedRobinDict("a"=>1, "b"=>2, "c"=>3);
+
+julia> get!(d, "a", 5)
+1
+
+julia> get!(d, "d", 4)
+4
+
+julia> d
+OrderedRobinDict{String,Int64} with 4 entries:
+  "a" => 1
+  "b" => 2
+  "c" => 3
+  "d" => 4
+```
+"""
+get!(collection, key, default)
+
+function get!(h::OrderedRobinDict{K,V}, key0, default) where {K,V}
+    key = convert(K,key0)
+
+    index = get(h.dict, key, -2)
+
+    index > 0 && return h.vals[index]
+
+    v = convert(V,  default)
```

Comment:
What is with the white space?

```suggestion
    key = convert(K,key0)
    index = get(h.dict, key, -2)
    index > 0 && return h.vals[index]
    v = convert(V,  default)
```